src/main/java/frc/robot/systems/IntakeFSMSystem.java

```diff
@@ -0,0 +1,707 @@
+		intakeMotor.setNeutralMode(NeutralModeValue.Brake);
+
+		throughBore = new Encoder(1, 2);
+		throughBore.reset();
+
+		colorSensor = new ColorSensorV3(Port.kOnboard);
+		led = new LED();
+
+		slot0Configs.kS = Constants.MM_CONSTANT_S;
+		slot0Configs.kV = Constants.MM_CONSTANT_V;
+		slot0Configs.kA = Constants.MM_CONSTANT_A;
+
+		slot0Configs.kP = Constants.MM_CONSTANT_P;
+		slot0Configs.kI = Constants.MM_CONSTANT_I;
+		slot0Configs.kD = 0;
+
+		motionMagicConfigs.MotionMagicAcceleration = Constants.CONFIG_CONSTANT_A;
+		motionMagicConfigs.MotionMagicJerk = Constants.CONFIG_CONSTANT_J;
+
+		statusCode = indexerMotor.getConfigurator().apply(talonFXConfigs);
+		statusCode = intakeMotor.getConfigurator().apply(talonFXConfigs);
+
+		// Reset state machine
+		reset();
+	}
+
+	/* ======================== Public methods ======================== */
+	/**
+	 * Return current FSM state.
+	 * @return Current FSM state
+	 */
+	public IntakeFSMState getCurrentState() {
+		return currentState;
+	}
+	/**
+	 * Reset this system to its start state. This may be called from mode init
+	 * when the robot is enabled.
+	 *
+	 * Note this is distinct from the one-time initialization in the constructor
+	 * as it may be called multiple times in a boot cycle,
+	 * Ex. if the robot is enabled, disabled, then reenabled.
+	 */
+	public void reset() {
+		led.greenLight(false);
+		currentState = IntakeFSMState.MOVE_TO_HOME;
+		hasNote = false;
+
+		timer.stop();
+		timer.reset();
+		// Call one tick of update to ensure outputs reflect start state
+		update(null);
+	}
+
+	/**
+	 * Update FSM based on new inputs. This function only calls the FSM state
+	 * specific handlers.
+	 * @param input Global TeleopInput if robot in teleop mode or null if
+	 *        the robot is in autonomous mode.
+	 */
+	public void update(TeleopInput input) {
+		switch (currentState) {
+			case MOVE_TO_HOME:
+				handleMoveHomeState(input);
+				break;
+
+			case MOVE_TO_GROUND:
+				handleMoveGroundState(input);
+				break;
+
+			case INTAKING:
+				handleIntakingState(input);
+				break;
+
+			case OUTTAKING:
+				handleOuttakingState(input);
+				break;
+
+			case FEED_SHOOTER:
+				handleFeedShooterState(input);
+				break;
+
+			default:
+				throw new IllegalStateException("Invalid state: " + currentState.toString());
+		}
+		currentState = nextState(input);
+
+		SmartDashboard.putString("CURRENT STATE", currentState.toString());
+		SmartDashboard.putNumber("Back Indexer Velocity",
+			indexerMotor.getVelocity().getValueAsDouble());
+		// SmartDashboard.putNumber("Front Indexer Velocity",
+		// 	frontIndexMotor.getVelocity().getValueAsDouble());
+		SmartDashboard.putBoolean("HASNOTE", hasNote);
+
+		SmartDashboard.putBoolean("Color Sensor in Range?", colorSensor.getProximity()
+			>= Constants.PROXIMIIY_THRESHOLD);
+		SmartDashboard.putNumber("Frames with Note in View", noteColorFrames);
+		SmartDashboard.putNumber("PIVOT ENCODER VAL", throughBore.getDistance());
+
+		hasNote = hasNote();
```

Comment:
Shouldn't you calculate this before putting in the dashboard? Also hasNote can be a function local variable, doesn't need to be a class variable.

Reply:
> Shouldn't you calculate this before putting in the dashboard?

Yes I will update that with the next commit
>Also hasNote can be a function local variable, doesn't need to be a class variable.

The reason it is a class variable is so that we don't need to call the hasNote() function in every method that requires it and instead we just set it equal to the hasNote variable every loop of the update() function and call the variable instead for all other purposes. I think we started doing this after MBR because part of the sensor code had an overrun issue and we believe it was the hasNote() function being called repeatedly. @beranki can speak more on that though